spec/syntax.md

```diff
@@ -0,0 +1,598 @@
+        _ [You have {$count} notifications.]
+
+A message with a single selector which is an invocation of
+a custom function `platform`, formatted on a single line:
+
+    {:platform} windows [Settings] _ [Preferences]
+
+A message with a single selector and a custom `hasCase` function
+which allows the message to query for presence of grammatical cases required for each variant:
+
+    {$userName: hasCase}
+        vocative [Hello, {$userName: person case=vocative}!]
+        accusative [Please welcome {$userName: person case=accusative}!]
+        _ [Hello!]
+
+A message with 2 selectors:
+
+    {$photoCount: number} {$userGender: equals}
+        one masculine [{$userName} added a new photo to his album.]
+        one feminine [{$userName} added a new photo to her album.]
+        one _ [{$userName} added a new photo to their album.]
+        _ masculine [{$userName} added {$photoCount} photos to his album.]
+        _ feminine [{$userName} added {$photoCount} photos to her album.]
+        _ _ [{$userName} added {$photoCount} photos to their album.]
+
+### Local Variables
+
+A message defining a local variable `$whom` which is then used twice inside the pattern:
+
+    $whom = {$monster: noun case=accusative}
+    [You see {$quality: adjective article=indefinite accord=$whom} {$whom}!]
+
+A message defining two local variables:
+`$itemAcc` and `$countInt`, and using `$countInt` as a selector:
+
+    $countInt = {$count: number maximumFractionDigits=0}
+    $itemAcc = {$item: noun count=$count case=accusative}
+        one [You bought {$color: adjective article=indefinite accord=$itemAcc} {$itemAcc}.]
+        _ [You bought {$countInt} {$color: adjective accord=$itemAcc} {$itemAcc}.]
+
+### Complex Messages
+
+A complex message with 2 selectors and 3 local variable definitions:
+
+    {$host: gender}
+    {$guestOther: number}
+
+    $hostName = {$host: person firstName=long}
+    $guestName = {$guest: person firstName=long}
+    $guestsOther = {$guestCount: number offset=1}
+
+        female 0 [{$hostName} does not give a party.]
+        female 1 [{$hostName} invites {$guestName} to her party.]
+        female 2 [{$hostName} invites {$guestName} and one other person to her party.]
+        female _ [{$hostName} invites {$guestName} and {$guestsOther} other people to her party.]
+
+        male 0 [{$hostName} does not give a party.]
+        male 1 [{$hostName} invites {$guestName} to his party.]
+        male 2 [{$hostName} invites {$guestName} and one other person to his party.]
+        male _ [{$hostName} invites {$guestName} and {$guestsOther} other people to his party.]
+
+        _ 0 [{$hostName} does not give a party.]
+        _ 1 [{$hostName} invites {$guestName} to their party.]
+        _ 2 [{$hostName} invites {$guestName} and one other person to their party.]
+        _ _ [{$hostName} invites {$guestName} and {$guestsOther} other people to their party.]
+
+## Comparison with ICU MessageFormat 1.0
```

Comment:
I think we can remove this whole section.

The differences should be visible from the grammar and examples.

Or at least add this section later, once the syntax is a bit more "established".
Otherwise we would have to keep updating this section every time we change the syntax.

Reply:
I think keeping this would be good, as it makes reading the spec easier for someone who may be familiar with MF1. As is, the PR as a whole is optimised for reading rather than editing, and it includes other places as well where the same grammar is expressed in multiple ways. Yes, that does mean that changes will need to be reflected in more than one place, but I would posit that that's a bearable cost when considering the benefits.

Reply:
OK, your call.

But I think that the differences are so much bigger in the general approach that comparing syntax details is irrelevant.

We will have custom formatter and custom selector functions, a standard function registry, better support for grammatical features, standard mapping to XLIFF, and this lists the `$`, `:`, and comma.

Reply:
Mihai has a point here. We can write a short comparison document (comparing more than just the syntax) once the design is more fleshed out. As of today, I think it's a bit too early and it will add maintenance burden for us.